File: app/src/main/java/pl/rpieja/flat/authentication/AccountService.java
package pl.rpieja.flat.authentication;

import android.accounts.AbstractAccountAuthenticator;
import android.accounts.Account;
import android.accounts.AccountAuthenticatorResponse;
import android.accounts.AccountManager;
import android.accounts.NetworkErrorException;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.IBinder;

/**
 * Created by maxmati on 11/18/17.
 */

public class AccountService extends Service {
    private Authenticator authenticator;
    public static final String ACCOUNT_TYPE = "pl.memleak.flat";

    public static void removeCurrentAccount(Context context) {
        AccountManager am = AccountManager.get(context);
        Account[] accounts = am.getAccountsByType(ACCOUNT_TYPE);
        if (accounts.length == 0) return;
        am.removeAccountExplicitly(accounts[0]);
    }
    public static void addAccount(Context context, String username, String token) {
        AccountManager am = AccountManager.get(context);
        Account account = AccountService.GetAccount(username);
        am.addAccountExplicitly(account, token, null);
    }

    public static String getAuthToken(Context context) {
        AccountManager am = AccountManager.get(context);
        Account[] accounts = am.getAccountsByType(ACCOUNT_TYPE);
        if(accounts.length == 0) return null;
        return am.getPassword(accounts[0]);
    }

    private static Account GetAccount(String accountName) {
        return new Account(accountName, ACCOUNT_TYPE);
    }

    @Override
    public void onCreate() {
        authenticator = new Authenticator(this);
    }

    @Override
    public IBinder onBind(Intent intent) {
        return authenticator.getIBinder();
    }

    private static class Authenticator extends AbstractAccountAuthenticator {

        Authenticator(Context context) {
            super(context);
        }

        @Override
        public Bundle editProperties(AccountAuthenticatorResponse accountAuthenticatorResponse, String s) {
            throw new UnsupportedOperationException();
        }

        @Override
        public Bundle addAccount(AccountAuthenticatorResponse accountAuthenticatorResponse, String s, String s1, String[] strings, Bundle bundle) throws NetworkErrorException {
            throw new UnsupportedOperationException();
        }

        @Override
        public Bundle confirmCredentials(AccountAuthenticatorResponse accountAuthenticatorResponse, Account account, Bundle bundle) throws NetworkErrorException {
            throw new UnsupportedOperationException();
        }

        @Override
        public Bundle getAuthToken(AccountAuthenticatorResponse accountAuthenticatorResponse, Account account, String s, Bundle bundle) throws NetworkErrorException {
            throw new UnsupportedOperationException();
        }

        @Override
        public String getAuthTokenLabel(String s) {
            throw new UnsupportedOperationException();
        }

        @Override
        public Bundle updateCredentials(AccountAuthenticatorResponse accountAuthenticatorResponse, Account account, String s, Bundle bundle) throws NetworkErrorException {
            throw new UnsupportedOperationException();
        }

        @Override
        public Bundle hasFeatures(AccountAuthenticatorResponse accountAuthenticatorResponse, Account account, String[] strings) throws NetworkErrorException {
            throw new UnsupportedOperationException();
        }
    }
}

File: app/src/main/java/pl/rpieja/flat/authentication/AccountService.kt
package pl.rpieja.flat.authentication

import android.accounts.AbstractAccountAuthenticator
import android.accounts.Account
import android.accounts.AccountAuthenticatorResponse
import android.accounts.AccountManager
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.IBinder
import pl.rpieja.flat.R
import pl.rpieja.flat.api.FlatAPI

class AccountService : Service() {
    private var authenticator: Authenticator? = null

    override fun onCreate() {
        authenticator = Authenticator(this)
    }

    override fun onBind(intent: Intent): IBinder? {
        return authenticator!!.iBinder
    }

    private class Authenticator internal constructor(context: Context) : AbstractAccountAuthenticator(context) {

        override fun editProperties(accountAuthenticatorResponse: AccountAuthenticatorResponse, s: String): Bundle {
            throw UnsupportedOperationException()
        }

        override fun addAccount(accountAuthenticatorResponse: AccountAuthenticatorResponse, s: String, s1: String, strings: Array<String>, bundle: Bundle): Bundle {
            throw UnsupportedOperationException()
        }

        override fun confirmCredentials(accountAuthenticatorResponse: AccountAuthenticatorResponse, account: Account, bundle: Bundle): Bundle {
            throw UnsupportedOperationException()
        }

        override fun getAuthToken(accountAuthenticatorResponse: AccountAuthenticatorResponse, account: Account, s: String, bundle: Bundle): Bundle {
            throw UnsupportedOperationException()
        }

        override fun getAuthTokenLabel(s: String): String {
            throw UnsupportedOperationException()
        }

        override fun updateCredentials(accountAuthenticatorResponse: AccountAuthenticatorResponse, account: Account, s: String, bundle: Bundle): Bundle {
            throw UnsupportedOperationException()
        }

        override fun hasFeatures(accountAuthenticatorResponse: AccountAuthenticatorResponse, account: Account, strings: Array<String>): Bundle {
            throw UnsupportedOperationException()
        }
    }

    companion object {
        private fun getAccountType(context: Context): String = context.getString(R.string.account_type)

        fun removeCurrentAccount(context: Context) {
            val am = AccountManager.get(context)
            val accounts = am.getAccountsByType(getAccountType(context))
            if (accounts.isEmpty()) return
            am.removeAccountExplicitly(accounts[0])
            FlatAPI.reset()
        }

        fun addAccount(context: Context, username: String, token: String) {
            val am = AccountManager.get(context)
            val account = AccountService.getAccount(context, username)
            am.addAccountExplicitly(account, token, null)
        }

        fun getAuthToken(context: Context): String? {
            val am = AccountManager.get(context)
            val accounts = am.getAccountsByType(getAccountType(context))
            return if (accounts.isEmpty()) null else am.getPassword(accounts[0])
        }

        private fun getAccount(context: Context, accountName: String): Account {
            return Account(accountName, getAccountType(context))
        }
    }
}
